### PR TITLE
remove resource from TF state when not found in OpsLevel during read

### PR DESCRIPTION
**Pull request by davidbloss**
Resolves [#508](https://github.com/OpsLevel/team-platform/issues/508)

### Problem

Terraform managed resources can also be deleted in the UI. When this happens the provider throws an error instead of updating the state to reflect reality.

### Solution

TF plans to remove resource when OpsLevel API specifically reports the resource does not exist

### Checklist

- [ ] I have run this code, and it appears to resolve the stated issue.
- [ ] This PR does not reduce total test coverage
- [ ] This PR has no user interface changes or has already received approval from product management to change the interface.
- [ ] Make a [changie](https://github.com/OpsLevel/terraform-provider-opslevel/blob/main/CONTRIBUTING.md#changie-change-log-generation) entry that explains the customer facing outcome of this change

### CODE REVIEW

File: opslevel/resource_opslevel_repository.go [hunk 1]
```diff
@@ -94,6 +94,10 @@ func (r *RepositoryResource) Create(ctx context.Context, req resource.CreateRequ
 		repository, err = r.client.GetRepositoryWithAlias(identifier)
 	}
 	if err != nil {
+		if (repository == nil || repository.Id == "") && opslevel.IsOpsLevelApiError(err) {
```
Comment (rocktavious) [inline opslevel/resource_opslevel_repository.go:97]: This is in "create" i'm not sure this is what we want to do

File: opslevel/resource_opslevel_team_tag.go [hunk 2]
```diff
@@ -162,12 +162,20 @@ func (teamTagResource *TeamTagResource) Read(ctx context.Context, req resource.R
...
 		resp.Diagnostics.AddError("opslevel client error", fmt.Sprintf("unable to read team (%s), got error: %s", teamIdentifier, err))
 		return
 	}
 	_, err = team.GetTags(teamTagResource.client, nil)
-	if err != nil || team.Tags == nil {
+	if err != nil {
+		if team.Tags == nil && opslevel.IsOpsLevelApiError(err) {
+			resp.State.RemoveResource(ctx)
+			return
+		}
```
Comment (rocktavious) [inline opslevel/resource_opslevel_team_tag.go:178]: I think this is another edgecase.  I think we want to error if the team is not found but RemoveResource if the tag is not found

File: opslevel/resource_opslevel_tag.go [hunk 3]
```diff
@@ -149,6 +149,10 @@ func (r *TagResource) Read(ctx context.Context, req resource.ReadRequest, resp *
 	resourceType := opslevel.TaggableResource(planModel.TargetType.ValueString())
 	data, err := r.client.GetTaggableResource(resourceType, resourceId)
 	if err != nil {
+		if (data == nil || data.ResourceId() == "") && opslevel.IsOpsLevelApiError(err) {
```
Comment (rocktavious) [inline opslevel/resource_opslevel_tag.go:152]: Same as the others - if the "resource" the tag should be on is not found thats an error - were as if the tag itself is not found thats `RemoveResource`

File: opslevel/resource_opslevel_service_tool.go [hunk 4]
```diff
@@ -140,6 +140,10 @@ func (r *ServiceToolResource) Create(ctx context.Context, req resource.CreateReq
 		service, err = r.client.GetServiceWithAlias(planModel.ServiceAlias.ValueString())
 	}
 	if err != nil {
+		if (service == nil || service.Id == "") && opslevel.IsOpsLevelApiError(err) {
```
Comment (rocktavious) [inline opslevel/resource_opslevel_service_tool.go:143]: Another one in create that we don't want.

File: opslevel/resource_opslevel_service_tag.go [hunk 5]
```diff
@@ -155,12 +155,20 @@ func (serviceTagResource *ServiceTagResource) Read(ctx context.Context, req reso
 		serviceIdentifier = data.ServiceAlias.ValueString()
 		service, err = serviceTagResource.client.GetServiceWithAlias(serviceIdentifier)
 	}
-	if err != nil || service == nil {
+	if err != nil {
```
Comment (rocktavious) [inline opslevel/resource_opslevel_service_tag.go:158]: Same as others where its a resource that depends on another resource being found first.

File: opslevel/resource_opslevel_service_repository.go [hunk 6]
```diff
@@ -203,6 +203,10 @@ func (r *ServiceRepositoryResource) Read(ctx context.Context, req resource.ReadR
 		service, err = r.client.GetServiceWithAlias(currentStateModel.ServiceAlias.ValueString())
 	}
 	if err != nil {
+		if (service == nil || service.Id == "") && opslevel.IsOpsLevelApiError(err) {
```
Comment (rocktavious) [inline opslevel/resource_opslevel_service_repository.go:206]: same issues - only `RemoveResource` if the `ServiceRepository` is not found on the service - error if the service isn't found.

File: opslevel/resource_opslevel_secrets.go [hunk 7]
```diff
@@ -131,6 +131,10 @@ func (r *SecretResource) Read(ctx context.Context, req resource.ReadRequest, res
 
 	secret, err := r.client.GetSecret(data.Id.ValueString())
 	if err != nil {
+		if (secret == nil || secret.ID == "") && opslevel.IsOpsLevelApiError(err) {
```
Comment (rocktavious) [inline opslevel/resource_opslevel_secrets.go:134]: This one feels scary - we need to ensure this one is something we want to correct.

File: opslevel/resource_opslevel_property_assignment.go [hunk 8]
```diff
@@ -139,11 +139,12 @@ func (resource *PropertyAssignmentResource) Read(ctx context.Context, req resour
 	owner := planModel.Owner.ValueString()
 	assignment, err := resource.client.GetProperty(owner, definition)
 	if err != nil {
+		if (assignment == nil || assignment.Definition.Id == "") && opslevel.IsOpsLevelApiError(err) {
```
Comment (rocktavious) [inline opslevel/resource_opslevel_property_assignment.go:142]: propery assignments don't have ID's - they are an assign / upsert kind of API - we don't need this.  In fact we should probably fully remove this error checking.

File: opslevel/resource_opslevel_secrets.go (hunk 7 above)
Comment (davidbloss) [reply opslevel/resource_opslevel_secrets.go:134]: How about adding a warning here like `resp.Diagnostics.AddWarning("State drift", fmt.Sprintf("Secret with ID '%s' in Terraform state not found in OpsLevel", data.Id.ValueString()))` ?

Comment (rocktavious) [reply opslevel/resource_opslevel_secrets.go:134]: Ok so thinking on it more i think its correct.  If the secret is not found by the ID then the worst it does is repush the secret to us.  Its not actually deleting anything.

File: opslevel/resource_opslevel_service_tool.go [hunk 9]
```diff
@@ -151,7 +151,11 @@ func (r *ServiceToolResource) Create(ctx context.Context, req resource.CreateReq
 		ServiceId:   &service.Id,
 		Url:         planModel.Url.ValueString(),
 	})
-	if err != nil || serviceTool == nil || string(serviceTool.Id) == "" {
+	if err != nil {
```
Comment (rocktavious) [inline opslevel/resource_opslevel_service_tool.go:154]: Should be in read?

**Review by rocktavious — APPROVED** [review]

**Pull request merged**